### PR TITLE
Range analysis and useless-comparison query: don't treat all unicode surrogates as if they are U+FFFD

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

**Comment by aschackmull** [comment]
I don't think this is the right fix. Instead I think that `CharacterLiteral.getCodePointValue` should be fixed to return the value as if the Java code had cast the character to an `int`. That's what the qldoc states, and I believe that was the intention of the predicate. We'll want something that uses the current implementation as much as possible, but for chars that would yield the result 65533, we should instead parse the literal.

**Comment by aschackmull** [comment]
The predicate `getCodePointValue` is fairly new, so I doubt there's any code relying on its current buggy behaviour.

**Comment by smowton** [comment]
:+1: pushed the in-place fix instead

File: java/ql/lib/semmle/code/java/Expr.qll [hunk 1]
```diff
@@ -731,7 +743,11 @@ class CharacterLiteral extends Literal, @characterliteral {
    * this literal. The result is the same as if the Java code had cast
    * the character to an `int`.
    */
-  int getCodePointValue() { result.toUnicode() = this.getValue() }
+  int getCodePointValue() {
+    if this.getLiteral().matches("'\\u%'")
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Expr.qll:747]: Let's make the match pattern slightly more precise.
```suggestion
    if this.getLiteral().matches("'\\u____'")
```

File: java/ql/lib/semmle/code/java/Expr.qll [hunk 2]
```diff
@@ -713,6 +713,18 @@ class DoubleLiteral extends Literal, @doubleliteral {
   override string getAPrimaryQlClass() { result = "DoubleLiteral" }
 }
 
+// Implementation taken from @p0 at https://github.com/github/codeql/issues/4145
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Expr.qll:716]: No need for this comment, I think.

**Comment by smowton** [comment]
@aschackmull done

**Review by aschackmull — APPROVED** [review]
It's technically possible to construct examples where `getCodePointValue` still returns the wrong value 65533 (when a surrogate literal doesn't match `"'\\u____'"`), but these examples are extremely obscure, so I'm not sure that it's worth the effort to properly support them until we see an actual need.

**Pull request merged**